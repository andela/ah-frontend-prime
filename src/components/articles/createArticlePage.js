import React, { Component } from "react";
import { connect } from "react-redux";
import "../../styles/app.scss";
import { articleCreateEditAction } from "../../actions/articleCreateEditAction";
import CreateArticleComponent from "./createArticleComponent";
import firebase from "../../firebase/config";

export class CreateArticlePage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: "",
      body: "",
      description: "",
      tags: "",
      image: "",
      isUploaded: false
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onUpload = this.onUpload.bind(this);
  }

  /* istanbul ignore next */
  onUpload(files) {
    const image = URL.createObjectURL(event.target.files[0]);
    this.setState({ image });

    const task = firebase
      .storage()
      .ref(`images/${files[0].name}`)
      .put(files[0]);

    task.then(res => {
      firebase
        .storage()
        .ref(`images/${files[0].name}`)
        .getDownloadURL()
        .then(url => {
          console.log(url);
          this.setState({ isUploaded: true, image: url });
        });
    });

    task.on("state_changed", snapshot => {
      const isUploaded = true;
      this.setState({ isUploaded: isUploaded });
    });
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }
  onSubmit(e) {
    e.preventDefault();
    const { title, body, description, tags, image } = this.state;
    const messageObject = {
      title: title,
      body: body,
      description: description,
      tags: tags.split(","),
      image: image
    };

    this.props.articleCreateEditAction(
      messageObject,
      "https://ah-backend-prime-staging.herokuapp.com/api/v1/articles/",
      "post",
      this.props
    );
  }

  render() {
    const { image } = this.state;
    return (
      <div>
        <CreateArticleComponent
          onChange={this.onChange}
          onUpload={this.onUpload}
          onSubmit={this.onSubmit}
          image={image}
        />
      </div>
    );
  }
}

export default connect(
  null,
  { articleCreateEditAction }
)(CreateArticlePage);
