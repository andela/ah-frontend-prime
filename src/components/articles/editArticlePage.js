import React, { Component } from "react";
import { connect } from "react-redux";
import "../../styles/app.scss";
import { articleCreateEditAction } from "../../actions/articleCreateEditAction";
import { getArticleAction } from "../../actions/getArticle";
import CreateArticleComponent from "./createArticleComponent";
import firebase from "../../firebase/config";
import { TextArea } from "semantic-ui-react";

export class EditArticlePage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: "",
      body: "",
      description: "",
      tags: "",
      image: "",
      isUploaded: false
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onUpload = this.onUpload.bind(this);
    this.onBodyChange = this.onBodyChange.bind(this);
  }
  componentWillMount() {
    this.props.getArticleAction(this.props.match.params.slug);
  }

  componentWillReceiveProps(newProps) {
    if (newProps) {
      this.setState({
        title: newProps.article.article.title,
        description: newProps.article.article.description,
        body: newProps.article.article.body,
        image: newProps.article.article.image,
        tags: newProps.article.tagList,
        article: newProps.article.article
      });
    }
  }

  /* istanbul ignore next */
  onUpload(files) {
    const image1 = URL.createObjectURL(event.target.files[0]);
    this.setState({ image1 });

    const task = firebase
      .storage()
      .ref(`images/${files[0].name}`)
      .put(files[0]);

    task.then(res => {
      firebase
        .storage()
        .ref(`images/${files[0].name}`)
        .getDownloadURL()
        .then(url => {
          this.setState({ isUploaded: true, image: url });
        });
    });

    task.on("state_changed", snapshot => {
      const isUploaded = true;
      this.setState({ isUploaded: isUploaded });
    });
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }
  onBodyChange(e) {
    this.setState({ body: e });
  }

  onSubmit(e) {
    e.preventDefault();
    const { title, body, description, tags, image } = this.state;
    const messageObject = {
      title: title,
      body: body,
      description: description,
      tags: tags ? tags.split(",") : "",
      image: image
    };
    const url =
      "https://ah-backend-prime-staging.herokuapp.com/api/v1/articles/" +
      this.props.match.params.slug +
      "/";

    this.props.articleCreateEditAction(
      messageObject,
      url,
      "put",
      this.props,
      "updated"
    );
  }

  render() {
    const { image, article } = this.state;

    return (
      <div>
        <CreateArticleComponent
          {...this.state}
          onChange={this.onChange}
          onUpload={this.onUpload}
          onSubmit={this.onSubmit}
          onBodyChange={this.onBodyChange}
          image={image}
          article={article ? article : null}
        />
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  article: state.getArticleReducer.article
});

export default connect(
  mapStateToProps,
  { getArticleAction, articleCreateEditAction }
)(EditArticlePage);
