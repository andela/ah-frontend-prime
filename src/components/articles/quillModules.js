export const modules = {};
export const formats = [];
