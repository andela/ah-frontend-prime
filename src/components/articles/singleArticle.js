import React, { Component } from "react";
import { connect } from "react-redux";
import moment from "moment";
import { Link } from "react-router-dom";

import "../../styles/singleArticle.scss";
import { getArticleAction } from "../../actions/getArticle";

export class SingleArticleComponent extends Component {
  componentDidMount() {
    this.props.getArticleAction(this.props.match.params.slug);
  }

  render() {
    const { article } = this.props;
    const oneArticle = article ? article.article : null;
    const singleArticle = oneArticle ? (
      <div className="container">
        <div className="single-article-div">
          <div className="inner">
            <h2 className="article-title">{oneArticle.title}</h2>
            <div>
              <div>
                <div className="">
                  <div className="article-profile">
                    <img
                      src={
                        oneArticle.author.image
                          ? oneArticle.author.image
                          : "https://static.wixstatic.com/media/2cd43b_4ef9dc9638cc431b956e1c36862b519b~mv2.png?dn="
                      }
                      alt="profile pic"
                    />
                  </div>
                  <div className="article-author">
                    {oneArticle.author.username}
                    <button className="btn btn-primary follow">follow</button>

                    <div className="article-time-stamps">
                      <div>
                        Created at{" "}
                        {moment(oneArticle.createdAt).format(
                          "h:mm:ss a [on] MMMM Do YYYY."
                        )}
                      </div>
                      <div>
                        modified at{" "}
                        {moment(oneArticle.updatedAt).format(
                          "h:mm:ss a [on] MMMM Do YYYY."
                        )}
                        .
                      </div>
                      <div>{oneArticle.reading_time}</div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className="article-single">
          <div className="article-description-body">
            <p>{oneArticle.description}</p>
          </div>
          <div className="image-div">
            <img
              src={
                oneArticle.image
                  ? oneArticle.image
                  : "https://previews.123rf.com/images/twindesigner/twindesigner1708/twin" +
                    "designer170800135/84202763-ah-brush-letter-logo-design-with-black-" +
                    "circle-creative-brushed-letters-icon-logo-.jpg"
              }
              className="article-image"
            />
          </div>

          <div className="article-description-body">
            <p style={{ fontWeight: "light" }}>{oneArticle.body}</p>
          </div>
        </div>
      </div>
    ) : (
      <div>The article you have requested does not exist</div>
    );
    return <div>{singleArticle}</div>;
  }
}

export const mapStateToProps = state => ({
  article: state.getArticleReducer.article
});

export default connect(
  mapStateToProps,
  { getArticleAction }
)(SingleArticleComponent);
