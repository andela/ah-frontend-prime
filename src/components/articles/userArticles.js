import React, { Component } from "react";
import { connect } from "react-redux";
import "../../styles/app.scss";
import moment from "moment";
import { withRouter } from "react-router-dom";
import { getArticlesAction } from "../../actions/getArticles";
import { deleteArticleAction } from "../../actions/deleteArticleAction";
import { Link } from "react-router-dom";

export class UserArticles extends Component {
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    this.props.getArticlesAction(
      "https://ah-backend-prime-staging.herokuapp.com/api/v1/articles/" +
        "?email=" +
        sessionStorage.getItem("email")
    );
  }
  deleteArticle(slug) {
    this.props.deleteArticleAction(slug, this.props);
  }
  render() {
    const articles = this.props.articles;
    const articleList =
      articles && articles.length ? (
        articles.map(article => {
          return (
            <div className="article" key={article.slug}>
              <div className="inner">
                <Link to={"/article/" + article.slug}>
                  <div className="article-card">
                    <h3>{article.title.substring(0, 30)} ...</h3>
                    <p>
                      <b>{article.description.substring(0, 60)} ...</b>
                    </p>
                    <p>{article.body.substring(0, 60)} ...</p>
                    <span>
                      <b>Author: {article.author.username}</b>
                    </span>
                    <br />
                    <span className="card-date">
                      Created at{" "}
                      {moment(article.createdAt).format(
                        "HH:MM:SS [on] MMMM Do YYYY."
                      )}
                    </span>
                  </div>
                  <div className="img-div">
                    <img
                      src={
                        article.image
                          ? article.image
                          : "https://previews.123rf.com/images/twindesigner/twindesigner1708/twin" +
                            "designer170800135/84202763-ah-brush-letter-logo-design-with-black-" +
                            "circle-creative-brushed-letters-icon-logo-.jpg"
                      }
                      className="img-tile"
                    />
                  </div>
                </Link>
                <div className="action-buttons">
                  <Link
                    to={{
                      pathname: "/edit-article/" + article.slug
                    }}
                  >
                    Edit
                  </Link>
                  <button
                    className="delete"
                    onClick={e => this.deleteArticle(article.slug)}
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          );
        })
      ) : (
        <div>You have not yet created any articles sofar</div>
      );
    return (
      <div className="big-container">
        <div className="user-articles">
          <div className="heading">
            <h3>{sessionStorage.getItem("username")}'s articles</h3>
          </div>
          <hr />
          {articleList}
        </div>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  articles: state.getArticlesReducer.articles
});

export default withRouter(
  connect(
    mapStateToProps,
    { getArticlesAction, deleteArticleAction }
  )(UserArticles)
);
