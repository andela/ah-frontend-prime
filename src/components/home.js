import React, { Component } from "react";
import { connect } from "react-redux";
import "../styles/app.scss";
import moment from "moment";
import { Link } from "react-router-dom";
import { getArticlesAction } from "../actions/getArticles";

export class Home extends Component {
  constructor(props) {
    super(props);
    this.fetchNext = this.fetchNext.bind(this);
    this.fetchPrevious = this.fetchPrevious.bind(this);
  }
  componentDidMount() {
    const { getArticlesAction } = this.props;
    getArticlesAction();
  }
  fetchNext() {
    const { next, getArticlesAction } = this.props;
    getArticlesAction(next);
  }
  fetchPrevious() {
    const { previous, getArticlesAction } = this.props;
    getArticlesAction(previous);
  }
  render() {
    const { articles } = this.props;
    const articleList = articles.length ? (
      articles.map(article => {
        return (
          <div className="article" key={article.slug}>
            <div className="inner">
              <Link
                className="links"
                to={{
                  pathname: "/article/" + article.slug
                }}
              >
                <div className="article-card">
                  <h3>{article.title.substring(0, 30)} ...</h3>
                  <p>
                    <b>{article.description.substring(0, 60)} ...</b>
                  </p>
                  <span>
                    <b>Author: {article.author.username}</b>
                  </span>
                  <br />

                  <span className="card-date">
                    Created at{" "}
                    {moment(article.createdAt).format(
                      "h:mm:ss a [on] MMMM Do YYYY."
                    )}
                  </span>
                  <br />
                  <p>
                    <b>{article.reading_time}</b>
                  </p>
                </div>

                <div>
                  <img
                    src={
                      article.image
                        ? article.image
                        : "https://previews.123rf.com/images/twindesigner/twindesigner1708/twin" +
                          "designer170800135/84202763-ah-brush-letter-logo-design-with-black-" +
                          "circle-creative-brushed-letters-icon-logo-.jpg"
                    }
                    className="img-tile"
                  />
                </div>
              </Link>
            </div>
          </div>
        );
      })
    ) : (
      <div>No Articles in Authors Haven Yet</div>
    );

    return (
      <div className="big-container">
        <div className="articles-top">
          <div className="best-article">
            <div className="inner">
              <img
                src={
                  "http://allpicts.in/download/22916/2018/03/Natural_Images_HD_1" +
                  "080p_Download_with_Waterfall_in_Tollymore_Forest_Park-1440x900.jpg/"
                }
                className="img-top"
              />
              <div className="inner">
                <div className="inner">
                  <div className="best-article-card">
                    <h3>
                      Facebook's Plan to Fuse its Messaging Apps Is Not about
                      Your Privacy Just follow the Money
                    </h3>
                    <span>
                      <b>Author: Yasha Levin in OneZero</b>
                    </span>
                    <br />
                    <span className="card-date">Created a few hours ago</span>
                  </div>
                </div>
              </div>
            </div>
            <div />
          </div>
          <div className="top-articles">{articleList}</div>
          <div className="top-articles">{articleList}</div>
        </div>

        <div className="articles">
          <div className="heading">
            <h3>Featured Articles</h3>
          </div>
          <hr />
          {articleList}
        </div>
        <div className="articles1">
          <div className="heading">
            <h3>Popular Articles</h3>
          </div>
          <hr />
          {articleList}
        </div>
        <div>
          <button
            className="btn btn-outline-success btn eian"
            onClick={() => this.fetchPrevious()}
          >
            Previous
          </button>{" "}
          <button
            className="btn btn-outline-success btn aken"
            onClick={() => this.fetchNext()}
          >
            Next
          </button>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  articles: state.getArticlesReducer.articles,
  next: state.getArticlesReducer.next,
  previous: state.getArticlesReducer.previous
});

export default connect(
  mapStateToProps,
  { getArticlesAction }
)(Home);
