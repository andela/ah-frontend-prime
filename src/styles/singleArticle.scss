.article-title {
  font-size: 40px;
  text-align: left;
  color: #000000;
}

.btn.follow {
  background: none;
  color: black;
  border: #17fb13 1px solid;
  margin-left: 5px;
  font-size: 12px;
  padding: 0 13px 0 13px;
}

.btn.follow:hover {
  background: none;
  color: black;
  border: #17fb13 1px solid;
  margin-left: 5px;
}

.article-profile > img {
  float: left;
  height: 50px;
  width: 50px;
  border: 1px solid #000000;
  border-radius: 50%;
  margin: 20px;
}

.article-image {
  display: block;
  max-width: 100%;
  max-height: 500px;
  margin-left: auto;
  margin-right: auto;
}

.article-author {
  float: left;
  width: 80%;
  margin-left: 20px;
}

.article-single {
  float: left;
  width: 100%;
  display: block;
}

.article-description-body {
  margin: 10px;
}

.image-div {
  width: 100%;
  background: rgba(0, 0, 0, 0.1);
}

.single-article-div {
  width: 100%;
}

form {
  .comments {
  }
}
