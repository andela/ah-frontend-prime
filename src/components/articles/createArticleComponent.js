import React from "react";
import "../../styles/createArticles.scss";
import "../../styles/singleArticle.scss";
import ReactQuill from "react-quill";
import { modules, formats } from "./quillModules";
import "react-quill/dist/quill.snow.css";
import "react-quill/dist/quill.bubble.css";

const CreateArticleComponent = props => {
  const { onChange, onSubmit, onUpload, onBodyChange, image, article, body } = props;

  return (
    <div className="container-fluid mt-5">
      <div className="row">
        <div className="col col-lg-3" />
        <div className="col col-lg-6">
          <h1>Tell A story</h1>
          <form>
            <div className="form-group">
              <input
                onChange={onChange}
                className="form-control"
                placeholder="title"
                name="title"
                id="title"
                defaultValue={article ? article.title : ""}
              />
            </div>
            <div className="form-group">
              <input
                onChange={onChange}
                className="form-control"
                placeholder="description"
                name="description"
                id="description"
                defaultValue={article ? article.description : ""}
              />
            </div>
            <div className="form-group">
              <ReactQuill
                onChange={onBodyChange}
                className="text-area"
                placeholder="Body"
                name="body"
                value={body}
                formats={formats}
                modules={modules}
              />
            </div>
            <div className="form-group">
              <input
                onChange={onChange}
                className="form-control"
                placeholder="tags"
                name="tags"
                id="tags"
                defaultValue={
                  article
                    ? article.tagList
                      ? article.tagList.join(",")
                      : ""
                    : ""
                }
              />
              <div>
                <i>Separate your tags with a comma</i>
              </div>
            </div>
            <div className="image-div">
              <div className="actual-image">
                <img
                  src={image ? image : article ? article.image : " "}
                  className="image-fluid"
                />
              </div>
              <input
                type="file"
                name="imageUpload"
                onChange={event => onUpload(event.target.files)}
                id="fitz"
              />
            </div>

            <div className="form-group">
              <button onClick={onSubmit} className="btn btn-primary">
                Publish Story
              </button>
            </div>
          </form>
        </div>
        <div className="col col-lg-3" />
      </div>
    </div>
  );
};

export default CreateArticleComponent;
