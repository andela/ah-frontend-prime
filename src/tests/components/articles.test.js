import React from "react";
import { Home } from "../../components/home";

import { shallow } from "enzyme";

import data from "../mock_data/moxios_mock";
import { SingleArticleComponent } from "../../components/articles/singleArticle";
import CreateArticleComponent from "../../components/articles/createArticleComponent";
import { CreateArticlePage } from "../../components/articles/createArticlePage";

const props = {
  articles: data.articles,
  previous: "",
  next: "",
  getArticlesAction: jest.fn(),
  fetchNext: jest.fn(),
  fetchPrevious: jest.fn()
};

const props1 = {
  getArticleAction: jest.fn(),
  match: {
    params: {
      slug: "this-is-kev"
    }
  }
};

describe("Article Components", () => {
  let homeWrapper;
  beforeEach(() => {
    homeWrapper = shallow(<Home {...props} />);
  });
  it("should render the home page", () => {
    expect(homeWrapper).toMatchSnapshot();
  });

  it("renders the single article page", () => {
    const wrapper = shallow(<SingleArticleComponent {...props1} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the single article page without an article", () => {
    props1.article = null;
    const wrapper = shallow(<SingleArticleComponent {...props1} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the single article page without an article image", () => {
    props1.article = data.articles[0];
    props1.article.image = null;
    const wrapper = shallow(<SingleArticleComponent {...props1} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should not regress", () => {
    const event = {
      target: {
        files: ["../../styles/images/profile.png", "sample.png"]
      }
    };
    const props = {
      onChange: jest.fn(),
      onSubmit: jest.fn(),
      onUpload: jest.fn()
    };

    const wrapper = shallow(<CreateArticleComponent {...props} />);
    wrapper.find("#fitz").simulate("change", event);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the create article page", () => {
    const wrapper = shallow(<CreateArticlePage />);
    wrapper.setProps({ article: "<div> we rocked so well</div>" });

    expect(wrapper).toMatchSnapshot();
  });

  it("should handle on change", () => {
    const event = {
      preventDefault: () => {},
      target: {
        name: "body",
        value: "This is the body"
      }
    };
    const props = {
      articleCreateEditAction: jest.fn()
    };
    const wrapper = shallow(<CreateArticlePage {...props} />);
    wrapper.instance().onChange(event);
    wrapper.instance().onSubmit(event);

    expect(wrapper.instance().state.body).toEqual("This is the body");
  });
  it("should fetch next page", () => {
    homeWrapper.instance().fetchNext();
    expect(props.getArticlesAction).toHaveBeenCalled();
  });
  it("should fetch previous page", () => {
    homeWrapper.instance().fetchPrevious();
    expect(props.getArticlesAction).toHaveBeenCalled();
  });
  it("should simulate onclick for next page", () => {
    homeWrapper
      .find(".aken")
      .at(0)
      .simulate("click");

    expect(props.fetchNext).toHaveBeenCalledTimes(0);
  });
  it("should simulate onclick for previous page", () => {
    homeWrapper
      .find(".eian")
      .at(0)
      .simulate("click");

    expect(props.fetchPrevious).toHaveBeenCalledTimes(0);
  });
});
