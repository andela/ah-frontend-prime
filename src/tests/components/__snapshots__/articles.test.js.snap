// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Components renders the create article page 1`] = `ShallowWrapper {}`;

exports[`Article Components renders the edit article page 1`] = `ShallowWrapper {}`;

exports[`Article Components renders the single article page 1`] = `ShallowWrapper {}`;

exports[`Article Components renders the single article page without an article 1`] = `ShallowWrapper {}`;

exports[`Article Components renders the single article page without an article image 1`] = `ShallowWrapper {}`;

exports[`Article Components should not regress 1`] = `ShallowWrapper {}`;

exports[`Article Components should render the home page 1`] = `ShallowWrapper {}`;
